protocol/opcm-bytecode-verification.md

```diff
@@ -0,0 +1,168 @@
+are set incorrectly).
+
+We want to eliminate this risk by automating the [Contract's Release
+Verification](https://www.notion.so/oplabs/Contracts-Release-Checklist-1f8f153ee162805e8236f022ebb8c868?source=copy_linkhttps:/)
+process, making it easy to demonstrate that an OPCM at a given address corresponds to a trusted
+commit.
+
+The solution should be:
+
+- contained within the release commit in the monorepo
+- easily runnable locally with a single command accepting only an RPC URL and the address of the
+  OPCM
+- runnable in CI in the `superchain-registry` repo
+- incorporated into the upgrade process in a way that ensures it is run by multiple people
+
+## Proposed Solution
+
+### Bytecode verification against the local source
+
+A new command should be added to `op-deployer`.
+
+```
+op-deployer verify-bytecode <opcm-address>
+
+```
+
+This command will invoke the `VerifyOPCM.s.sol` script's [default
+entrypoint](https://github.com/ethereum-optimism/optimism/blob/158e990b76a85acbb018577bd4079190b2d97281/packages/contracts-bedrock/scripts/deploy/VerifyOPCM.s.sol#L126-L129)
+to verify the OPCM.
+
+```
+op-deployer verify-bytecode --single-contract <contract-name> <contract-address>
+```
+
+This command will invoke the `VerifyOPCM.s.sol` script's [runSingle
+entrypoint](https://github.com/ethereum-optimism/optimism/blob/158e990b76a85acbb018577bd4079190b2d97281/packages/contracts-bedrock/scripts/deploy/VerifyOPCM.s.sol#L135)
+in order to verify any contracts involved in the upgrade which are not included in the OPCM (ie. the
+new `DeputyPauseModule` introduced in upgrade 16).
+
+### Artifacts source
+
+By default, `op-deployer verify-bytecode` will use locally built forge-artifacts to check bytecode.
+In order to facilitate quickly running in CI, without having to checkout and rebuild different
+commits, the command will also accept a tag locator, with the following invocation:
+
+```
+op-deployer verify-bytecode --dangerously-use-remote-artifacts --artifacts-locator tag://op-contracts/vX.Y.Z
+
+```
+
+The flag `--dangerously-use-remote-artifacts` is intended to discourge the use of remote artifacts
+when running locally, while still enabling a fast mechanism to run in CI.
+
+### OPCM config verification
+
+`op-deployer verify-bytecode` will require one of the following groups of arguments to confirm the
+configuration of the OPCM.
+
+Address flags:
+
+```
+--upgrade-controller 0x... \
+  --superchain-config 0x... \
+  --protocol-versions 0x... \
+  --superchain-proxy-admin 0x...
+```
+
+Superchain-registry flag:
+
+```
+--superchain-registry <path/to/registry>
+```
+
+The argument groups should be mutually exclusive, either all of the address flags or the
+superchain-registry flag should be provided, but not both.
+
+### Integration points
+
+It is important to ensure that the bytecode verification process is run by multiple people and on
+multiple different machines.
```

Comment:
Why is it important for multiple people to run this? What are the risks if only one person runs the verification?